在performance結束後offset tool鏈接超時無法顯示資訊

![Screenshot from 2022-02-20 23-34-37](https://user-images.githubusercontent.com/43372856/154850771-7c50ea98-367c-4e17-835e-c0fe796c21c0.png)
![Screenshot from 2022-02-20 23-33-08](https://user-images.githubusercontent.com/43372856/154850777-04c37aae-ac2d-477d-bedb-1d1f5fd3e888.png)

如圖，能夠通過kafka tool獲取topic name，但我無法用offset獲取分布，它會顯示超時。在performance中是可以獲取到分布的，上述情況爲performance結束後。
可能還需要確認一下上述情況是否必會處罰該bug。

```
> Task :app:TopicExplorer.main()
[main] INFO org.apache.kafka.clients.admin.AdminClientConfig - AdminClientConfig values: 
	bootstrap.servers = [0.0.0.0:11262, 0.0.0.0:15548]
	client.dns.lookup = use_all_dns_ips
	client.id = 
	connections.max.idle.ms = 300000
	default.api.timeout.ms = 60000
	metadata.max.age.ms = 300000
	metric.reporters = []
	metrics.num.samples = 2
	metrics.recording.level = INFO
	metrics.sample.window.ms = 30000
	receive.buffer.bytes = 65536
	reconnect.backoff.max.ms = 1000
	reconnect.backoff.ms = 50
	request.timeout.ms = 30000
	retries = 2147483647
	retry.backoff.ms = 100
	sasl.client.callback.handler.class = null
	sasl.jaas.config = null
	sasl.kerberos.kinit.cmd = /usr/bin/kinit
	sasl.kerberos.min.time.before.relogin = 60000
	sasl.kerberos.service.name = null
	sasl.kerberos.ticket.renew.jitter = 0.05
	sasl.kerberos.ticket.renew.window.factor = 0.8
	sasl.login.callback.handler.class = null
	sasl.login.class = null
	sasl.login.connect.timeout.ms = null
	sasl.login.read.timeout.ms = null
	sasl.login.refresh.buffer.seconds = 300
	sasl.login.refresh.min.period.seconds = 60
	sasl.login.refresh.window.factor = 0.8
	sasl.login.refresh.window.jitter = 0.05
	sasl.login.retry.backoff.max.ms = 10000
	sasl.login.retry.backoff.ms = 100
	sasl.mechanism = GSSAPI
	sasl.oauthbearer.clock.skew.seconds = 30
	sasl.oauthbearer.expected.audience = null
	sasl.oauthbearer.expected.issuer = null
	sasl.oauthbearer.jwks.endpoint.refresh.ms = 3600000
	sasl.oauthbearer.jwks.endpoint.retry.backoff.max.ms = 10000
	sasl.oauthbearer.jwks.endpoint.retry.backoff.ms = 100
	sasl.oauthbearer.jwks.endpoint.url = null
	sasl.oauthbearer.scope.claim.name = scope
	sasl.oauthbearer.sub.claim.name = sub
	sasl.oauthbearer.token.endpoint.url = null
	security.protocol = PLAINTEXT
	security.providers = null
	send.buffer.bytes = 131072
	socket.connection.setup.timeout.max.ms = 30000
	socket.connection.setup.timeout.ms = 10000
	ssl.cipher.suites = null
	ssl.enabled.protocols = [TLSv1.2, TLSv1.3]
	ssl.endpoint.identification.algorithm = https
	ssl.engine.factory.class = null
	ssl.key.password = null
	ssl.keymanager.algorithm = SunX509
	ssl.keystore.certificate.chain = null
	ssl.keystore.key = null
	ssl.keystore.location = null
	ssl.keystore.password = null
	ssl.keystore.type = JKS
	ssl.protocol = TLSv1.3
	ssl.provider = null
	ssl.secure.random.implementation = null
	ssl.trustmanager.algorithm = PKIX
	ssl.truststore.certificates = null
	ssl.truststore.location = null
	ssl.truststore.password = null
	ssl.truststore.type = JKS

[main] INFO org.apache.kafka.common.utils.AppInfoParser - Kafka version: 3.1.0
[main] INFO org.apache.kafka.common.utils.AppInfoParser - Kafka commitId: 37edeed0777bacb3
[main] INFO org.apache.kafka.common.utils.AppInfoParser - Kafka startTimeMs: 1645804733539
[kafka-admin-client-thread | adminclient-1] INFO org.apache.kafka.clients.NetworkClient - [AdminClient clientId=adminclient-1] Node -2 disconnected.
[kafka-admin-client-thread | adminclient-1] WARN org.apache.kafka.clients.NetworkClient - [AdminClient clientId=adminclient-1] Connection to node -2 (/127.0.0.1:15548) could not be established. Broker may not be available.
[kafka-admin-client-thread | adminclient-1] INFO org.apache.kafka.common.utils.AppInfoParser - App info kafka.admin.client for adminclient-1 unregistered
[kafka-admin-client-thread | adminclient-1] INFO org.apache.kafka.common.metrics.Metrics - Metrics scheduler closed
[kafka-admin-client-thread | adminclient-1] INFO org.apache.kafka.common.metrics.Metrics - Closing reporter org.apache.kafka.common.metrics.JmxReporter
[kafka-admin-client-thread | adminclient-1] INFO org.apache.kafka.common.metrics.Metrics - Metrics reporters closed
Exception in thread "main" java.lang.RuntimeException: java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.TimeoutException: Timed out waiting for a node assignment. Call: metadata
	at org.astraea.Utils.handleException(Utils.java:24)
	at org.astraea.topic.Builder$TopicAdminImpl.earliestOffset(Builder.java:131)
	at org.astraea.topic.Builder$TopicAdminImpl.offsets(Builder.java:213)
	at org.astraea.topic.TopicExplorer.execute(TopicExplorer.java:65)
	at org.astraea.topic.TopicExplorer.main(TopicExplorer.java:105)
Caused by: java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.TimeoutException: Timed out waiting for a node assignment. Call: metadata
Caused by: java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.TimeoutException: Timed out waiting for a node assignment. Call: metadata

	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2069)
	at org.apache.kafka.common.internals.KafkaFutureImpl.get(KafkaFutureImpl.java:165)
	at org.astraea.topic.Builder$TopicAdminImpl.lambda$earliestOffset$17(Builder.java:138)
	at org.astraea.Utils.handleException(Utils.java:22)
	... 4 more
Caused by: org.apache.kafka.common.errors.TimeoutException: Timed out waiting for a node assignment. Call: metadata
Caused by: org.apache.kafka.common.errors.TimeoutException: Timed out waiting for a node assignment. Call: metadata


> Task :app:TopicExplorer.main() FAILED

Execution failed for task ':app:TopicExplorer.main()'.
> Process 'command '/usr/lib/jvm/java-16-jdk/bin/java'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.



```

--- a/app/src/main/java/org/astraea/topic/TopicExplorer.java
+++ b/app/src/main/java/org/astraea/topic/TopicExplorer.java
@@ -73,8 +73,8 @@ public class TopicExplorer {
                     .thenComparing(TopicPartition::partition))
             .collect(Collectors.toUnmodifiableList());
     if (!invalidTopics.isEmpty()) {
-      // The AdminClient#listOffsets API call will time out if any of the requested partitions
-      // has no leader yet.
+      // The metadata request sent by Kafka admin succeeds only if the leaders of all partitions are
+      // alive. Hence, we throw exception here to make consistent behavior with Kafka.
       throw new IllegalStateException("Some partitions have no leader: " + invalidTopics);
     }
 

--- a/app/src/main/java/org/astraea/topic/TopicExplorer.java
+++ b/app/src/main/java/org/astraea/topic/TopicExplorer.java
@@ -63,6 +63,18 @@ public class TopicExplorer {
 
   static Result execute(TopicAdmin admin, Set<String> topics) {
     var replicas = admin.replicas(topics);
+
+    var invalidTopics =
+        replicas.entrySet().stream()
+            .filter(x -> x.getValue().stream().noneMatch(Replica::leader))
+            .map(Map.Entry::getKey)
+            .collect(Collectors.toUnmodifiableList());
+    if (!invalidTopics.isEmpty()) {
+      // The AdminClient#listOffsets API call will time out if any of the requested partitions
+      // has no leader yet.
+      throw new IllegalStateException("Some partitions have no leader: " + invalidTopics);
+    }
+
     var offsets = admin.offsets(topics);
     var consumerGroups = admin.consumerGroup(Set.of());
     var time = LocalDateTime.now();

--- a/app/src/main/java/org/astraea/topic/TopicExplorer.java
+++ b/app/src/main/java/org/astraea/topic/TopicExplorer.java
@@ -68,6 +68,9 @@ public class TopicExplorer {
         replicas.entrySet().stream()
             .filter(x -> x.getValue().stream().noneMatch(Replica::leader))
             .map(Map.Entry::getKey)
+            .sorted(
+                Comparator.comparing(TopicPartition::topic)
+                    .thenComparing(TopicPartition::partition))
             .collect(Collectors.toUnmodifiableList());
     if (!invalidTopics.isEmpty()) {
       // The AdminClient#listOffsets API call will time out if any of the requested partitions

--- a/app/src/test/java/org/astraea/topic/TopicExplorerTest.java
+++ b/app/src/test/java/org/astraea/topic/TopicExplorerTest.java
@@ -1,6 +1,7 @@
 package org.astraea.topic;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
 import java.io.ByteArrayOutputStream;
@@ -23,6 +24,7 @@ import org.astraea.consumer.Consumer;
 import org.astraea.service.RequireBrokerCluster;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
 
 public class TopicExplorerTest extends RequireBrokerCluster {
 
@@ -200,5 +202,31 @@ public class TopicExplorerTest extends RequireBrokerCluster {
     assertTrue(output.matches("(?ms).+\\[.*non-synced.*].+"), "descriptor printed");
     assertTrue(output.matches("(?ms).+\\[.*future.*].+"), "descriptor printed");
     assertTrue(output.matches("(?ms).+/tmp/path0.+"), "path printed");
+  }
+
+  @Test
+  void somePartitionsHaveNoLeader() {
+    String topicName0 = "poor_topic0";
+    String topicName1 = "poor_topic1";
+    var payload0 =
+        Map.of(
+            new TopicPartition(topicName0, 0),
+            List.of(new Replica(1000, 0, 0, false, false, false, "?")));
+    var payload1 =
+        Map.of(
+            new TopicPartition(topicName1, 0),
+            List.of(
+                new Replica(1000, 0, 0, false, false, false, "?"),
+                new Replica(1001, 0, 0, false, false, false, "?"),
+                new Replica(1002, 0, 0, false, false, false, "?")));
+    TopicAdmin mock = Mockito.mock(TopicAdmin.class);
+
+    Mockito.when(mock.replicas(Set.of(topicName0))).thenReturn(payload0);
+    Mockito.when(mock.replicas(Set.of(topicName1))).thenReturn(payload1);
+
+    assertThrows(
+        IllegalStateException.class, () -> TopicExplorer.execute(mock, Set.of(topicName0)));
+    assertThrows(
+        IllegalStateException.class, () -> TopicExplorer.execute(mock, Set.of(topicName1)));
   }
 }

